src/include/daos_srv/bio.h

```diff
@@ -764,4 +772,48 @@ int bio_replace_dev(struct bio_xs_context *xs, uuid_t old_dev_id,
  */
 int bio_set_led_state(struct bio_xs_context *xs, uuid_t devid,
 		      const char *led_state, bool reset);
+
+/*
+ * Allocate DMA buffer, the buffer could be from bulk cache if bulk context
+ * if specified.
+ *
+ * \param ioctxt	[IN]	I/O context
+ * \param len		[IN]	Requested buffer length
+ * \param bulk_ctxt	[IN]	Bulk context
+ * \param bulk_perm	[IN]	Bulk permission
+ *
+ * \return			Buffer descriptor on success, NULL on error
+ */
+struct bio_desc *bio_buf_alloc(struct bio_io_context *ioctxt,
+			       unsigned int len, void *bulk_ctxt,
+			       unsigned int bulk_perm);
+
+/*
+ * Free allocated DMA buffer.
+ *
+ * \param biod		[IN]	Buffer descriptor
+ *
+ * \return			N/A
+ */
+void bio_buf_free(struct bio_desc *biod);
+
+/*
+ * Get the bulk handle of DMA buffer.
+ *
+ * \param biod		[IN]	Buffer descriptor
+ * \param bulk_off	[OUT]	Bulk offset
+ *
+ * \return			Bulk handle
+ */
+void *bio_buf_bulk(struct bio_desc *biod, unsigned int *bulk_off);
+
+/*
+ * Get the address of DMA buffer.
+ *
+ * \param biod		[IN]	Buffer descriptor
+ *
+ * \return			Buffer address
+ */
+void *bio_buf_addr(struct bio_desc *biod);
```

Comment:
Since we already have the bulk handle for the DMA buffer, why still need  the address?

Reply:
The bio_buf_alloc() is used to allocate a region of DMA buffer, the allocated region isn't necessary to have a associated cached bulk handle. Even if we have the cart bulk handle, we still needs some way to get the buffer address, right?